### DISCUSSION TITLE
Faster `ExplicitBitVect` to numpy array?

### DISCUSSION THREAD

**Discussion by hadim** [rdkit/rdkit#3863] [Q&A] [answered]
I recently realized there is a faster way to convert a rdkit fingerprint to a numpy array and I would like to know whether my approach is correct.

Consider the following fp:

```python
from rdkit import Chem
from rdkit.Chem import rdMolDescriptors
from rdkit.Chem import DataStructs

import numpy as np

mol = Chem.MolFromSmiles("CN1C=NC2=C1C(=O)N(C(=O)N2C)C")

fp_vec = rdMolDescriptors.GetMorganFingerprintAsBitVect(
    mol,
    radius=2,
    nBits=2048,
    useFeatures=True,
)
```

and the two below methods to convert it to a numpy array:

```python
%%timeit -n 1000 -r 1000
# 10.1 µs ± 3.62 µs per loop (mean ± std. dev. of 1000 runs, 1000 loops each)
np.frombuffer(fp_vec.ToBitString().encode(), 'u1') - ord('0')
```

```python
%%timeit -n 1000 -r 100
fp = np.zeros((0,), dtype=int)
# 39.6 µs ± 1.45 µs per loop (mean ± std. dev. of 100 runs, 1000 loops each)
DataStructs.ConvertToNumpyArray(fp_vec, fp)
```

Finally check the arrays are the same:

```python
fp1 = np.frombuffer(fp_vec.ToBitString().encode(), 'u1') - ord('0')

fp2 = np.zeros((0,), dtype=int)
DataStructs.ConvertToNumpyArray(fp_vec, fp2)

np.array_equal(fp1, fp2)
# True
```

You can notice a 4-time speedup using the `np.frombuffer` approach.

This benchmark is really quick, only tested with one molecule and on a single machine so I might be missing something here.

Test system: python 3.9.1 and rdkit 2020.09.4

**Comment by mwojcikowski** [comment]
Hi @hadim 

I had a "challenge accepted" moment :) The best alternative and a bit less hacky is the following:
```python
np.unpackbits(np.frombuffer(DataStructs.BitVectToBinaryText(fp_vec), dtype=np.uint8), bitorder='little')
```
But this is still slower than your approach (12us vs 10us).

I tried to profile a bit and it turned out that the numpy part is negligible and the real hog are the RDKit methods, and with your initial approach used the fastest RDKit method to serialize a FP amongst the ones I tried

![image](https://user-images.githubusercontent.com/2325645/109686599-b81ff880-7b82-11eb-99a2-c31237418678.png)

**Comment by kienerj** [comment]
According to numpy docs using `frombuffer()` results in a read-only array. However I guess with fingerprints that is a non-issue in most cases but probably explains why it isn't the default way of doing it.

**Reply by hadim** [reply]
The result of `frombuffer` is not returned instead we return `frombuffer() - <something>` so the returned array is regular and can be written:

```python
# Without `- ord('0')`
fp = np.frombuffer(fp_vec.ToBitString().encode(), 'u1')
# fp.flags["WRITEABLE"] -> False
fp[5:10] = 0
# raises: `ValueError: assignment destination is read-only`

# With `- ord('0')`
fp = np.frombuffer(fp_vec.ToBitString().encode(), 'u1') - ord('0')
# fp.flags["WRITEABLE"] -> True
fp[5:10] = 0
# works
```

**Comment by renzph** [comment]
I recently also tried to optimize this and came to the same solution. 

`np.frombuffer(bytes(fp_vec.ToBitString(), 'utf-8'), 'u1') - ord('0')`

That's faster than all the other things I tried. 
Thanks for sharing.

**Comment by IchiruTake** [comment]


**Reply by IchiruTake** [reply]
**UPDATE**:
I have run the method 02 above and see that for 50000 loops, at size 256, 1024, 1536, 2048, the time consumed is equivalent to around 2.25 to 2.5 seconds with vectorization.

**Reply by IchiruTake** [reply]
Update 02: 
When I deepening the source code, I found that the method: `bitVect.ToBitString()` actually points to `BitVectToText` in `rdkit/DataStructs/__init__.py` in `Code/DataStructs/BitOps.cpp`. Instead of writing function as above, the fastest way to construct the numpy would be something as (written in C++)

```
def BitVectToNumpyArray(bitVect: Union[ExplicitBitVect, SparseBitVect]):
      unsigned const int size = bitVect.getNumBits()
      value = np.uint8(1)
      ndarray arr = np.zeros((size , ), dtype=np.uint8)
      for (unsigned int i = 0; i < size ; i++) 
      { 
            if (bitVect.getBit(i)) {  arr[i] = value ; } 
      }
      return arr
```
I believe the above function can increase around 10% or more compared to `np.frombuffer(self.ToBitString().encode(), 'u1') - ord('0')`. Another option is using some cache with BitVect by another class but we disable change in this class such as ExplicitStrictBitVect and SparseStrictBitVect. The amount of cache would take up around 10% memory extra using list/tuple/ndarray for index storage. 

The function `StrictBitVectToNumpyArray` would have same implementation as `BitVectToNumpyArray`, but the function would be: 
```
def StrictBitVectToNumpyArray(bitVect: Union[ExplicitBitVect, SparseBitVect]):
      unsigned const int size = bitVect.getNumBits()
      value = np.uint8(1)
      indices = bitVect.getIndexOnBits()     //pointer: vector or indexes of unsigned integer stored position of on bits
      ndarray arr = np.zeros((size , ), dtype=np.uint8)
      arr[indices] = 1
      return arr
```

**Comment by greglandrum** [comment, marked as answer]
fwiw I have a PR open here: https://github.com/rdkit/rdkit/pull/4523 which adds the ability to get fingerprints directly as numpy objects
This should be in the 2021.09.1 release (coming in October)
